File: src/main/java/endytkn/randomEvents/events/SkeletonFightEvent.java
package endytkn.randomEvents.events;

import endytkn.randomEvents.baseEvents.groupFight.GroupFight;
import endytkn.randomEvents.baseEvents.groupFight.GroupFightBaseEvent;
import endytkn.randomEvents.randomEvent.RandomEvent;
import endytkn.randomEvents.randomEvent.RandomEventsCategory;
import endytkn.randomEvents.randomEvent.RandomEventsRarity;
import net.minecraft.network.chat.Component;
import net.minecraft.world.InteractionHand;
import net.minecraft.world.effect.MobEffectInstance;
import net.minecraft.world.effect.MobEffects;
import net.minecraft.world.entity.EntityType;
import net.minecraft.world.entity.EquipmentSlot;
import net.minecraft.world.entity.LightningBolt;
import net.minecraft.world.entity.Mob;
import net.minecraft.world.item.ItemStack;
import net.minecraft.world.item.Items;
import net.minecraft.world.item.enchantment.Enchantments;

import java.util.*;

public class SkeletonFightEvent extends GroupFightBaseEvent {
    public SkeletonFightEvent() {
        super(true);
        categories = List.of(RandomEventsCategory.GROUP_FIGHT, RandomEventsCategory.DAY_NIGHT, RandomEventsCategory.WEATHER_CLEAR, RandomEventsCategory.WEATHER_THUNDER, RandomEventsCategory.WEATHER_RAIN);
        this.rarity = RandomEventsRarity.COMMON;
        this.eventTag = "skeleton_fight";
    }

    @Override
    public RandomEvent create() {
        return new SkeletonFightEvent();
    }

    @Override
    public void onPrepare() {
        LightningBolt lightningBolt = EntityType.LIGHTNING_BOLT.create(level);
        lightningBolt.setCause(null);
        lightningBolt.setPos(targetBlock.getX(), targetBlock.getY(), targetBlock.getZ());
        lightningBolt.setSecondsOnFire(0);
        level.addFreshEntity(lightningBolt);

        int minionsCount = 2 + playersGroup.size() * 2 + new Random().nextInt(2);
        int leaderCount = minionsCount / 3;

        Map<UUID, Mob> skeletonMobs = new HashMap<>();

        for (int i = 0; i < minionsCount; i++) {
            EntityType<? extends Mob> mobType = EntityType.SKELETON;
            Mob mob = mobType.create(level);
            this.addEntityToEvent(mob);
            if (mob != null) {
                ItemStack weapon = new ItemStack(i % 3 == 0 ? Items.BOW : Items.STONE_SWORD);
                mob.setCanPickUpLoot(false);
                mob.setItemInHand(InteractionHand.MAIN_HAND, weapon);
                skeletonMobs.put(mob.getUUID(), mob);
            }
        }

        for (int i = 0; i < leaderCount; i++) {
            EntityType<? extends Mob> mobType = EntityType.SKELETON;
            Mob mob = mobType.create(level);
            if (mob != null) {
                ItemStack weapon = new ItemStack(i % 3 == 0 ? Items.BOW : Items.GOLDEN_SWORD);
                mob.setHealth(mob.getHealth() * 2);
                mob.setCustomName(Component.literal("Skeleton Leader"));

                ItemStack helmet = new ItemStack(Items.LEATHER_HELMET);
                ItemStack chestplate = new ItemStack(Items.LEATHER_CHESTPLATE);
                helmet.enchant(Enchantments.UNBREAKING, 5);
                chestplate.enchant(Enchantments.UNBREAKING, 5);
                this.addEntityToEvent(mob);
                mob.addEffect(new MobEffectInstance(MobEffects.GLOWING, this.timeLimit));
                mob.setItemSlot(EquipmentSlot.HEAD, helmet);
                mob.setItemSlot(EquipmentSlot.CHEST, chestplate);
                mob.setCanPickUpLoot(false);
                mob.setItemInHand(InteractionHand.MAIN_HAND, weapon);
                skeletonMobs.put(mob.getUUID(), mob);
            }
        }

        GroupFight skeletonGroup = new GroupFight(this, "skeleton", skeletonMobs, null);
        mobGroupies.put(skeletonGroup.groupName, skeletonGroup);

        super.onPrepare();
    }
}

File: src/main/kotlin/endytkn/randomEvents/events/SkeletonFightEvent.kt
package endytkn.randomEvents.events

import endytkn.randomEvents.baseEvents.GroupFight.GroupFight
import endytkn.randomEvents.baseEvents.GroupFight.GroupFightBaseEvent
import endytkn.randomEvents.randomEvents.RandomEvent
import endytkn.randomEvents.randomEvents.RandomEventsCategories
import endytkn.randomEvents.randomEvents.RandomEventsRarity
import net.minecraft.network.chat.Component
import net.minecraft.world.InteractionHand
import net.minecraft.world.effect.MobEffectInstance
import net.minecraft.world.effect.MobEffects
import net.minecraft.world.entity.EntityType
import net.minecraft.world.entity.EquipmentSlot
import net.minecraft.world.entity.Mob
import net.minecraft.world.item.ItemStack
import net.minecraft.world.item.Items
import net.minecraft.world.item.enchantment.Enchantments
import java.util.UUID
import kotlin.random.Random

class SkeletonFightEvent : GroupFightBaseEvent(true) {
    companion object {
        val name: String = "skeleton_fight"
    }

    var minionsCount: Int? = null
    var leaderCount: Int? = null

    init {
        category = RandomEventsCategories.GROUP_FIGHT
        rarity = RandomEventsRarity.COMMON
        this.eventTag = "skeletonFight"
    }

    override fun create(): RandomEvent {
        return SkeletonFightEvent()
    }

    override fun onPrepare() {
        val lightningBolt = EntityType.LIGHTNING_BOLT.create(level!!)
        lightningBolt!!.setPos(targetBlock!!.x.toDouble(), targetBlock!!.y.toDouble(), targetBlock!!.z.toDouble())
        lightningBolt.setSecondsOnFire(0)
        level!!.addFreshEntity(lightningBolt)
        minionsCount = Random.nextInt(2 + this.playersGroup!!.size * 2, 4 + this.playersGroup!!.size * 2)
        leaderCount = minionsCount!! / 3
        val skeletonMobs = mutableMapOf<UUID, Mob>()

        repeat(minionsCount!!) { index ->
            var mobType: EntityType<out Mob> = EntityType.SKELETON
            var mob = mobType.create(level!!)
            mob?.let {
                val weapon = ItemStack(if (index % 3 == 0) Items.BOW else Items.STONE_SWORD)
                it.setCanPickUpLoot((false))
                it.setItemInHand(InteractionHand.MAIN_HAND, weapon)
                skeletonMobs.put(it.uuid, it)
            }
        }

        repeat(leaderCount!!) { index ->
            var mobType: EntityType<out Mob> = EntityType.SKELETON
            var mob = mobType.create(level!!)
            mob?.let {
                val weapon =  ItemStack(if (index % 3 == 0) Items.BOW else Items.GOLDEN_SWORD)
                it.health = (mob.health * 2)
                it.customName = Component.literal("Skeleton Leader")
                val helmet = ItemStack(Items.GOLDEN_HELMET)
                val chestplate = ItemStack(Items.GOLDEN_CHESTPLATE)
                helmet.enchant(Enchantments.UNBREAKING, 5)
                chestplate.enchant(Enchantments.UNBREAKING, 5)
                it.addEffect(MobEffectInstance(MobEffects.GLOWING, this.timeLimit))
                it.setItemSlot(EquipmentSlot.HEAD, helmet)
                it.setItemSlot(EquipmentSlot.CHEST, chestplate)
                it.setCanPickUpLoot((false))
                it.setItemInHand(InteractionHand.MAIN_HAND, weapon)
                skeletonMobs.put(it.uuid, it)
            }
        }
        val skeletonGroup = GroupFight(this, "skeleton", skeletonMobs, null)
        mobGroupies.put(skeletonGroup.groupName, skeletonGroup)
        super.onPrepare()
    }
}